modules/text2vec-openai/vectorizer/objects.go

```diff
@@ -87,3 +115,181 @@ func (v *Vectorizer) getVectorizationConfig(cfg moduletools.ClassConfig) ent.Vec
+				}
+				break
+			}
+			start := time.Now()
+			rateLimitNew, err := v.makeRequest(job, texts, conf, vecBatchOffset)
+			batchTookInS = time.Since(start).Seconds()
+			timePerToken = batchTookInS / float64(tokensInCurrentBatch)
+			if rateLimitNew != nil {
+				rateLimit = rateLimitNew
+			}
+			// not all request limits are included in "RemainingRequests" and "ResetRequests". For example, in the free
+			// tier only the RPD limits are shown but not RPM
+			if rateLimit.RemainingRequests == 0 {
+				// if we need to wait more than MaxBatchTime for a reset we need to stop the batch to not produce timeouts
+				if time.Since(batchStart)+time.Duration(rateLimit.ResetRequests)*time.Second > MaxBatchTime {
+					for j := vecBatchOffset; j < len(job.texts); j++ {
+						job.errs[j] = err
+					}
+					break
+				}
+				time.Sleep(time.Duration(rateLimit.ResetRequests) * time.Second)
+			}
+
+			// reset for next vectorizer-batch
+			vecBatchOffset = objCounter
+			tokensInCurrentBatch = 0
+			texts = texts[:0]
+		}
+		job.wg.Done()
+
+	}
+}
+
+func (v *Vectorizer) makeRequest(job batchJob, texts []string, conf ent.VectorizationConfig, vecBatchOffset int,
+) (*ent.RateLimits, error) {
+	res, rateLimit, err := v.client.Vectorize(job.ctx, texts, conf)
+	if err != nil {
+		for j := 0; j < len(texts); j++ {
+			job.errs[vecBatchOffset+j] = err
+		}
+	} else {
+		for j := 0; j < len(texts); j++ {
+			if res.Errors[j] != nil {
+				job.errs[vecBatchOffset+j] = res.Errors[j]
+			} else {
+				job.vecs[vecBatchOffset+j] = res.Vector[j]
+			}
+		}
+	}
+
+	return rateLimit, err
+}
+
+func (v *Vectorizer) ObjectBatch(ctx context.Context, objects []*models.Object, skipObject []bool, cfg moduletools.ClassConfig,
+) ([][]float32, map[int]error) {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	errs := make(map[int]error)
+	texts := make([]string, len(objects))
+	tokens := make([]int, len(objects))
+	conf := v.getVectorizationConfig(cfg)
+	icheck := NewClassSettings(cfg)
+	vecs := make([][]float32, len(objects))
+
+	// go token library is outdated. Alter the model-name to use a different model name with the same tokenization-behaviour
+	model := conf.Model
+	if model == "text-embedding-ada-002" || model == "text-embedding-3-small" || model == "text-embedding-3-large" {
+		model = "gpt-4"
+	}
+
+	tke, err := tiktoken.EncodingForModel(model)
+	if err != nil { // fail all objects as they all have the same model
+		for j := range objects {
+			errs[j] = err
+		}
+		return nil, errs
+	}
+
+	// prepare input for vectorizer, and send it to the queue. Prepare here to avoid work in the queue-worker
+	for i := range objects {
+		if skipObject[i] {
+			continue
+		}
+		text := v.objectVectorizer.Texts(ctx, objects[i], icheck)
+		texts[i] = text
+		tokens[i] = clients.GetTokensCount(conf.Model, text, tke)
+	}
+
+	v.jobQueueCh <- batchJob{ctx: ctx, wg: &wg, errs: errs, cfg: cfg, texts: texts, tokens: tokens, vecs: vecs, skipObject: skipObject}
```

Comment:
timer may be needed to skip adding job to queue that got expired and to return errors. seems like max batch time can be exceeded even before job will be pushed to the queue and handled by worker (if there are multiple requests waiting)

Reply:
note: if channel would be buffered then timer would have to mark jobs already added to the queue as expired and return errors. marked jobs can be then ignored by worker.

Reply:
I moved to a buffered channel, but as explained in another comment it is more a soft-limit than a hard-limit, so I don't want to kill off batches just based on that. Maaybe in the future, but as the current batch implementation also does not handle this case I would like to leave it as it is